index_alias_impl.go

```diff
@@ -428,19 +473,136 @@ func (i *indexAliasImpl) Swap(in, out []Index) {
+	return requestHasKNN(req)
+}
+
+func PreSearch(ctx context.Context, req *SearchRequest, indexes ...Index) ([]map[string]interface{}, error) {
+	// create a dummy request with a match none query
+	// since we only care about the preSearchData in PreSearch
+	dummyRequest := &SearchRequest{
+		Query: query.NewMatchNoneQuery(),
+	}
+	newCtx := context.WithValue(ctx, search.PreSearchKey, true)
+	if requestHasKNN(req) {
+		addKnnToDummyRequest(dummyRequest, req)
+	}
+	res, err := PreSearchDataSearch(newCtx, dummyRequest, indexes...)
+	if err != nil {
+		return nil, err
+	}
+	preSearchData, err := mergePreSearchData(req, res, len(indexes))
+	if err != nil {
+		return nil, err
+	}
+	return preSearchData, nil
+}
+
+func tagHitsWithIndexNum(sr *SearchResult, indexNum int) {
+	for _, hit := range sr.Hits {
+		hit.IndexId = append(hit.IndexId, indexNum)
+	}
+}
+
+func mergePreSearchData(req *SearchRequest, res *SearchResult, numIndexes int) ([]map[string]interface{}, error) {
+	mergedOut := make([]map[string]interface{}, numIndexes)
+	for i := 0; i < len(mergedOut); i++ {
+		mergedOut[i] = make(map[string]interface{})
+	}
+	if requestHasKNN(req) {
+		mergeKNNDocumentMatches(req, res.Hits, mergedOut)
+	}
+	return mergedOut, nil
+}
+
+func PreSearchDataSearch(ctx context.Context, req *SearchRequest, indexes ...Index) (*SearchResult, error) {
+	searchStart := time.Now()
+	asyncResults := make(chan *asyncSearchResult, len(indexes))
+
+	// run search on each index in separate go routine
+	var waitGroup sync.WaitGroup
+
+	var searchChildIndex = func(in Index, childReq *SearchRequest, idx int) {
+		rv := asyncSearchResult{Name: in.Name(), IndexNum: idx}
+		rv.Result, rv.Err = in.SearchInContext(ctx, childReq)
+		asyncResults <- &rv
+		waitGroup.Done()
+	}
+
+	waitGroup.Add(len(indexes))
+	for idx, in := range indexes {
+		go searchChildIndex(in, createChildSearchRequest(req, nil), idx)
+	}
+
+	// on another go routine, close after finished
+	go func() {
+		waitGroup.Wait()
+		close(asyncResults)
+	}()
+
+	var sr *SearchResult
+	indexErrors := make(map[string]error)
+
+	for asr := range asyncResults {
+		if asr.Err == nil {
+			if sr == nil {
+				// first result
+				sr = asr.Result
+				tagHitsWithIndexNum(sr, asr.IndexNum)
+			} else {
+				// merge with previous
+				tagHitsWithIndexNum(asr.Result, asr.IndexNum)
+				sr.Merge(asr.Result)
+			}
+		} else {
+			indexErrors[asr.Name] = asr.Err
+		}
+	}
+
+	// merge just concatenated all the hits
+	// now lets clean it up
+
+	// handle case where no results were successful
+	if sr == nil {
+		sr = &SearchResult{
+			Status: &SearchStatus{
+				Errors: make(map[string]error),
+			},
+		}
+	}
+
+	searchDuration := time.Since(searchStart)
+	sr.Took = searchDuration
+
+	// fix up errors
+	if len(indexErrors) > 0 {
+		if sr.Status.Errors == nil {
+			sr.Status.Errors = make(map[string]error)
+		}
+		for indexName, indexErr := range indexErrors {
+			sr.Status.Errors[indexName] = indexErr
+			sr.Status.Total++
```

Comment:
Why are we incrementing Total only if errors are greater than 0?

Reply:
As far as I see, on line 554, we aren't incrementing Total in 'sr', so am I missing something?

Reply:
Oh never mind, realised my oversight.
sr.Merge() does increment Total.